src/internal/memory.rs

```diff
@@ -53,27 +54,50 @@ impl<P: Package, V: Version> Memory<P, V> {
 
     /// Add a decision to a Memory.
```

Comment:
That check that was done in the removed `add_version` was important. It prevents generating a future conflict and thus avoid a full pass of conflict resolution. That's the heart of what I meant when I said that removing the `remove_decision` function would mean a few changes in the algorithm. We have to have a way to prevent that decision been added if one of it's dependencies incompatibilities would become satisfied.

One option I can see is checking if the dependencies are almost satisfied before adding the decision. If that's the case, it is necessarily because the unsatisfied term is the decision we want to add. And in that case we shouldn't add it since it would mean that we will satisfy one of those incompatibilities.

Reply:
> One option I can see is checking if the dependencies are almost satisfied before adding the decision

Actually this is wrong since the derivations intersection for the package we are adding could already match the decision term, and in such case the incompatibility could already be fully satisfied, not just almost satisfied. But since we've already validated the dependencies at that point, the correct (and most efficient) way is I think to use the dependency ranges directly. We could check that none of the `not(range)` terms for dependency ranges are satisfied by the partial solution before adding the decision. Because if any of them are satisfied, we know for sure that the incompatibility `{p: v, dep: not(range)}` will be satisfied once we add the decision for version `v`.

Reply:
> avoid a full pass of conflict resolution.

The full pass will end up in the same place, so that is what I had in mind by calling this a "fast path". But if it is important there are a number of "simple" ways to add it back in. One is to redefine `remove_last_decision` as `bactrack(self.decision -1)`. Another is to "ask permission not forgiveness" in this case, and... it is hard to say in words but it worked so take a look.

Also now I think we only need to do this if the dependencies are new, as otherwise the incompatibilities are already accounted for by the previous call to `unit_propagation`.